Support XML metadata files and modify DMD sections

jraddaoui opened the pull request:
Add support for XML metadata files including configurable validation
and addition to the METS file as DMD sections on ingest and re-ingest.
Modify DMD sections management to improve versioning and indexing.

Refs https://github.com/archivematica/Issues/issues/1531 and https://github.com/archivematica/Issues/issues/1537.

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -284,6 +285,104 @@ def createDMDIDsFromCSVMetadata(job, path, state):
+    #   - Consider equal filenames from different transfer sources?
+    #   - Allow multiple XML metadata per file?
+    #   - Use extra CSV to relate XML file(s)?
+    xmlFilename = os.path.splitext(os.path.relpath(path, "objects"))[0] + ".xml"
```
Could we start using [pathlib](https://docs.python.org/3.6/library/pathlib.html) for path operations and directory traversal? Also, now that Python 2 has been removed entirely from the latest `qa/1.x` I don't think we have to deal with import errors anymore.

jraddaoui replied:
Cool, a lot cleaner.

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -1141,6 +1240,16 @@ def createFileSec(
                         f.originallocation.replace("%transferDirectory%", "", 1),
                         state,
                     )
+                    XMLDMDID = createDMDFromXML(
```
Should this be renamed to `createDMDIDFromXML` to match the name pattern in the previous function above?

jraddaoui replied:
I had some conflicts about this ... I don't think that pattern represents the actual purpose of those functions. I also thought about renaming the other one and including this call inside it, since the params are the same in the end, but I don't want to mess a lot with existing functions.

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -1141,6 +1240,16 @@ def createFileSec(
+                    if DMDIDS and XMLDMDID:
+                        DMDIDS += " " + XMLDMDID
+                    elif XMLDMDID:
+                        DMDIDS = XMLDMDID
```
Would something like `DMDIDS = ' '.join(i for i in [DMDIDS, XMLDMDID] if i)` work to avoid the `if`?

jraddaoui replied:
Neat, thanks!

replaceafill commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -372,6 +372,7 @@ def _get_index_settings():
     return {
         "index": {
             "mapping": {
+                "ignore_malformed": True,
```
Will this require to recreate the index(es) somehow?

jraddaoui replied:
Yes, this is just a hack for now to avoid errors importing XMLs with different structure in the same elements.

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -1141,6 +1240,16 @@ def createFileSec(
                         f.originallocation.replace("%transferDirectory%", "", 1),
                         state,
                     )
+                    XMLDMDID = create_dmd_from_xml(
```
```suggestion
                    XMLDMDID = create_dmdid_from_xml(
```

jraddaoui replied:
Mentioned above.

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -284,6 +285,104 @@ def createDMDIDsFromCSVMetadata(job, path, state):
+    #   - Consider equal filenames from different transfer sources?
+    #   - Allow multiple XML metadata per file?
+    #   - Use extra CSV to relate XML file(s)?
+    xml_filename = os.path.splitext(os.path.relpath(path, "objects"))[0] + ".xml"
```
Could we use [pathlib](https://docs.python.org/3.6/library/pathlib.html) for path handling and directory traversal here? Now that Python 2 has been removed in the latest `qa/1.x` branch I don't think we even need to worry about [the old `ImportErrors`](https://github.com/artefactual/archivematica/blob/bf635157bb829124104b7d0d9278a579c61c261d/src/archivematicaCommon/lib/archivematicaFunctions.py#L39-L42).

jraddaoui replied:
Addressed above.

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -284,6 +285,104 @@ def createDMDIDsFromCSVMetadata(job, path, state):
+    )
+    if not os.path.isdir(transfer_metadata):
+        return
+    for dir in os.listdir(transfer_metadata):
```
Should we rename [`dir`](https://docs.python.org/3.6/library/functions.html#dir) since it's a built-in function?

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -284,6 +285,104 @@ def createDMDIDsFromCSVMetadata(job, path, state):
+            continue
+        try:
+            tree = etree.parse(xml_path)
+            valid, errors = _validate_xml(tree)
```
Would it be worth separating parsing and validation in the `try` block to provide more accurate logging?

jraddaoui replied:
Good point. Since validation has its own exception handling I've moved it out from this try.

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -284,6 +285,104 @@ def createDMDIDsFromCSVMetadata(job, path, state):
+            )
+            return
+        root = tree.getroot()
+        _, _, tag = root.tag.partition("}")
```
Could we use [`QName(root).localname`](https://lxml.de/api/lxml.etree.QName-class.html) to extract the tag without the namespace?

jraddaoui replied:
Nice, thanks!!

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -284,6 +285,104 @@ def createDMDIDsFromCSVMetadata(job, path, state):
+        root = tree.getroot()
+        _, _, tag = root.tag.partition("}")
+        state.globalDmdSecCounter += 1
+        DMDID = "dmdSec_" + state.globalDmdSecCounter.__str__()
```
```suggestion
        DMDID = "dmdSec_{}".format(state.globalDmdSecCounter)
```

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -284,6 +285,104 @@ def createDMDIDsFromCSVMetadata(job, path, state):
+    return True, []
+
+
+def _get_schema_uri(tree):
```
Can we copy the comment from the `common` setting module about this logic?

jraddaoui replied:
Sure, waiting a little bit more to see how much of this will be needed in the end, before adding docstrings and tests for these functions.

replaceafill reviewed (commented):
@jraddaoui I think this approach is convenient for our current knowledge of the use case and I understand it's still work in progress, but I just left a few comments with potential improvements.

jraddaoui commented:
Thanks for the great feedback @replaceafill, I think I've addressed it all (in commits or comments). Fixing the tests now.

I also added the initial implementation to get the XML metadata to original files mapping from a `source-metadata.csv` file. If you could take a quick look, that would be great.

replaceafill commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSMetadataXML.py:
```diff
@@ -0,0 +1,265 @@
+
+
+def process_xml_metadata(mets, sip_dir, sip_uuid, sip_type):
+    if not getattr(mcpclient_settings, "XML_VALIDATION", False):
```
Do you think it's worth moving this condition back to the [caller](https://github.com/artefactual/archivematica/blob/65c54429ba34fe43137c63e20e89b4a98dd48895/src/MCPClient/lib/clientScripts/create_mets_v2.py#L1890-L1893)?

jraddaoui replied:
Since we still need to pass that setting, I think it's okay to have it here.

replaceafill commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSMetadataXML.py:
```diff
@@ -0,0 +1,265 @@
+
+def _get_schema_uri(tree):
+    XSI = "http://www.w3.org/2001/XMLSchema-instance"
+    VALIDATION = mcpclient_settings.XML_VALIDATION
```
Do you think is worth trying to avoid the `mcpclient_settings` import and cascade this as a parameter from the caller to the top level function and to here?

replaceafill commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSMetadataXML.py:
```diff
@@ -0,0 +1,265 @@
+    key = None
+    checked_keys = []
+    schema_location = tree.xpath(
+        "/*/@xsi:noNamespaceSchemaLocation", namespaces={"xsi": XSI}
```
Could this reuse the [namespaces.NSMAP](https://github.com/artefactual/archivematica/blob/76f6faa1d895e3a83923b45c9f5a5b18cfab1a3e/src/archivematicaCommon/lib/namespaces.py#L26-L35)?

replaceafill commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSMetadataXML.py:
```diff
@@ -0,0 +1,265 @@
+def _add_validation_event(mets, file_uuid, schema_uri, valid, errors):
+    event_data = {
+        "eventType": "validation",
+        "eventDetail": 'type="metadata"; validation-source-type="'
+        + schema_uri.split(".")[-1]
+        + '"; validation-source="'
+        + schema_uri
+        + '"; program="lxml"; version="'
+        + etree.__version__
+        + '"',
```
This part is a bit hard to read. What do you think of splitting it like this:

```
    event_detail = {
        "type": "metadata",
        "validation-source-type": schema_uri.split(".")[-1],
        "validation-source": schema_uri,
        "program": "lxml",
        "version": etree.__version__,
    }
...
        "eventDetail": "; ".join(
            [f'{k}="{v}"' for k, v in event_detail.items()]
        ),
...
```

jraddaoui replied:
Nice, thanks!

replaceafill commented on src/MCPClient/lib/clientScripts/create_mets_v2.py:
```diff
@@ -36,12 +37,13 @@
 
 django.setup()
 # dashboard
-from django.utils import timezone
```
Why is `timezone.now()` not relevant anymore?

jraddaoui replied:
It was used to add the CREATEDATE attribute to the METS header. Now, with the inclusion of the metsrw on the final steps, if that attribute is set when the METS is parsed, metsrw adds the LASTMODDATE attribute thinking it's a re-ingest.

I removed that at some point and then introduced `datetime.datetime` for the new dates, to keep consistency with the metsrw.

replaceafill commented on src/MCPClient/tests/test_xml_metadata.py:
```diff
@@ -0,0 +1,570 @@
+#!/usr/bin/env python
```
Should this module be renamed to `tests_archivematicaCreateMETSMetadataXML` to hint what's covering?

replaceafill reviewed (approved):
Great work @jraddaoui! I left a couple of questions and suggestions, but this looks good to me.

jraddaoui committed:
Support XML metadata files and modify DMD sections

Add support for XML metadata files including configurable validation
and addition to the METS file as DMD sections on ingest and re-ingest.
Modify DMD sections management to improve versioning and indexing.
